### ISSUE TITLE
Getting an error when adding a product item to shopping cart

### ISSUE THREAD

**Issue by DrewBrasher** [OrchardCMS/OrchardCore.Commerce#73]
Getting a `Value cannot be null. (Parameter 'key')` exception when adding a product item to shopping cart.  This exception also occurs (silently if not debugging) when creating or viewing a product item.

**Steps to reproduce:**

1. Clone OrchardCore.Commerce
2. Run the SampleWebApp project
3. Complete setup (I used the Agency recipe and Sqlite)
4. Enable all 3 features under "Commerce": Orchard Core Commerce, Orchard Core Commerce Session Cart Storage, Orchard Core Commerce Settings Currency Selector
5. Run all three recipes under "Orchard Core Commerce": MultiCurrencyProduct, Order, Product
6. Add a "Product" from the "New" menu.  Visual Studio displays the exception below but if you continue past it you can still add the Product.
```
System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null. (Parameter 'key')
  Source=System.Private.CoreLib
  StackTrace:
   at System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Money.CurrencyProvider.GetCurrency(String isoSymbol) in C:\repos\OrchardCore.Commerce\MoneyDataType\CurrencyProvider.cs:line 20
   at Money.Currency.FromISOCode(String code, IEnumerable`1 providers) in C:\repos\OrchardCore.Commerce\MoneyDataType\Currency.extra.cs:line 137
   at OrchardCore.Commerce.Settings.CommerceSettingsCurrencySelector.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Settings\CommerceSettingsCurrencySelector.cs:line 16
   at OrchardCore.Commerce.Services.MoneyService.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\MoneyService.cs:line 52
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 64
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass3_0.<Edit>b__0(PricePartViewModel m) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 38
   at OrchardCore.DisplayManagement.Handlers.DisplayDriverBase.<>c__DisplayClass8_0`1.<Initialize>b__0(TModel shape) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Handlers\DisplayDriverBase.cs:line 62
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<CreateAsync>g__ShapeFactory|6_1[TModel](Func`2 init) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 87
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<CreateAsync>b__2() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 95
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.<CreateAsync>d__10.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 92
```
7. Fill out the Title, Body, and SKU fields and click "Publish"
8. Click "View" to view the product you just added.  Visual Studio displays the exception below but if you continue past it you can still view the Product.
```
System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null. (Parameter 'key')
  Source=System.Private.CoreLib
  StackTrace:
   at System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Money.CurrencyProvider.GetCurrency(String isoSymbol) in C:\repos\OrchardCore.Commerce\MoneyDataType\CurrencyProvider.cs:line 20
   at Money.Currency.FromISOCode(String code, IEnumerable`1 providers) in C:\repos\OrchardCore.Commerce\MoneyDataType\Currency.extra.cs:line 137
   at OrchardCore.Commerce.Settings.CommerceSettingsCurrencySelector.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Settings\CommerceSettingsCurrencySelector.cs:line 16
   at OrchardCore.Commerce.Services.MoneyService.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\MoneyService.cs:line 52
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 64
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass2_0.<Display>b__0(PricePartViewModel m) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 27
   at OrchardCore.DisplayManagement.Handlers.DisplayDriverBase.<>c__DisplayClass8_0`1.<Initialize>b__0(TModel shape) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Handlers\DisplayDriverBase.cs:line 62
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<CreateAsync>g__ShapeFactory|6_1[TModel](Func`2 init) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 87
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<CreateAsync>b__2() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 95
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.<CreateAsync>d__10.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 92
```
9. Click "Add to cart".  The browser displays the exception below
```
System.ArgumentNullException: Value cannot be null. (Parameter 'key')
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Money.CurrencyProvider.GetCurrency(String isoSymbol) in C:\repos\OrchardCore.Commerce\MoneyDataType\CurrencyProvider.cs:line 20
   at Money.Currency.FromISOCode(String code, IEnumerable`1 providers) in C:\repos\OrchardCore.Commerce\MoneyDataType\Currency.extra.cs:line 137
   at OrchardCore.Commerce.Settings.CommerceSettingsCurrencySelector.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Settings\CommerceSettingsCurrencySelector.cs:line 16
   at OrchardCore.Commerce.Services.MoneyService.get_CurrentDisplayCurrency() in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\MoneyService.cs:line 52
   at OrchardCore.Commerce.Services.PriceProvider.<>c__DisplayClass5_0.<AddPrices>b__3(PricePart pricePart) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\PriceProvider.cs:line 41
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at OrchardCore.Commerce.Models.ShoppingCartItem..ctor(Int32 quantity, String productSku, IEnumerable`1 attributes, IEnumerable`1 prices) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Models\ShoppingCartItem.cs:line 33
   at OrchardCore.Commerce.Models.ShoppingCartItem.WithPrices(IEnumerable`1 prices) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Models\ShoppingCartItem.cs:line 64
   at OrchardCore.Commerce.Services.PriceProvider.<>c__DisplayClass5_0.<AddPrices>b__1(ShoppingCartItem item) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\PriceProvider.cs:line 43
   at System.Linq.Utilities.<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.SelectArrayIterator`2.MoveNext()
   at System.Linq.Set`1.UnionWith(IEnumerable`1 other)
   at System.Linq.Enumerable.DistinctIterator`1.FillSet()
   at System.Linq.Enumerable.DistinctIterator`1.ToArray()
   at OrchardCore.Commerce.Services.PriceVariantProvider.AddPrices(IEnumerable`1 items) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\PriceVariantProvider.cs:line 29
   at OrchardCore.Commerce.Services.PriceService.AddPrices(IEnumerable`1 items) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Services\PriceService.cs:line 25
   at OrchardCore.Commerce.Controllers.ShoppingCartController.AddItem(ShoppingCartLineUpdateModel line, String shoppingCartId) in C:\repos\OrchardCore.Commerce\OrchardCore.Commerce\Controllers\ShoppingCartController.cs:line 105
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Diagnostics\DiagnosticsStartupFilter.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at OrchardCore.ContentPreview.PreviewStartupFilter.<>c.<<Configure>b__1_1>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.ContentPreview\PreviewStartupFilter.cs:line 23
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 63
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 236
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 59
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by bleroy** [comment]
Weird, I can't repro that.

**Comment by bleroy** [comment]
What currency have you been using on that product?

**Comment by DrewBrasher** [comment]
If I go to the admin side and go to "Content > Content Definition > Content Types" and edit the Product Content Type, then edit the Price under the Parts section, it shows "All Currencies".  I also tried changing it to "Specific Currency" and selected USD but I still get the same result.
![image](https://user-images.githubusercontent.com/9272802/99865845-2ad80b80-2b72-11eb-8a0f-1381725fbae9.png)

Maybe it's because I don't have a price on the Product Item I added but I don't see a way to add a price.  All I see when adding a new product item is Title, Permalink, HtmlBody, and SKU.
![image](https://user-images.githubusercontent.com/9272802/99865911-ad60cb00-2b72-11eb-8cb8-cabc581f2464.png)

**Comment by DAud-IcI** [comment]
This error you mentioned with the `Money.CurrencyProvider.GetCurrency(string isoCode)` has been fixed in a different issue [here](https://github.com/OrchardCMS/OrchardCore.Commerce/commit/9ade1a5d9a7f582310c4a6d2a86582c695fbd9b4#diff-1a510a37522d25342c57023632b4c9f0c2aca9d5b6c4432069a8d316670e752eR18).

There may be still other errors because the underlying problem is that the default currency ISO code is not set. This will be addressed in #105. Until that one is resolved the workaround is:
- go to **Admin** > **Configuration** > **Settings** > **Commerce**
- click on the <kbd>Save</kbd> button

This will reset your currency configuration to USD instead of null.
